oak_runtime/src/node/http/mod.rs

```diff
@@ -518,21 +494,122 @@ impl HttpResponseIterator {
                     .status(StatusCode::INTERNAL_SERVER_ERROR)
                     .body(Body::empty())
+                    .map_err(|err| {
+                        log::warn!("Could not build response: {}", err);
+                        OakStatus::ErrInternal
+                    })
             }
         }
     }
 }
+
+// Wrapper around a HashMap representing the headers in protobuf encapsulated HttpRequest and
+// HttpResponse. This struct is defined mainly to allow implementing `std::convert::From` for more
+// convenient conversion from hyper HeaderMaps.
+struct OakHeaderMap {
+    headers: HashMap<String, HeaderValue>,
+}
+
+impl OakHeaderMap {
+    // Flatten the headers into a vector of tuples (String, http::header::HeaderValue) that can
+    // be easily iterated over when populating the headers in the hyper Response builder.
+    fn to_vec(&self) -> Vec<(String, http::header::HeaderValue)> {
+        self.headers
+            .iter()
+            .flat_map(|item| {
+                item.1.values.iter().map(move |val| {
+                    let header_value = match http::header::HeaderValue::from_bytes(val) {
+                        Ok(header_value) => header_value,
+                        Err(err) => {
+                            log::warn!("Error when mapping response: {}", err);
+                            http::header::HeaderValue::from_static("")
+                        }
+                    };
+
+                    (item.0.to_owned(), header_value)
+                })
+            })
+            .collect()
+    }
+}
+
+impl From<&http::header::HeaderMap<http::header::HeaderValue>> for OakHeaderMap {
+    fn from(hyper_headers: &http::header::HeaderMap<http::header::HeaderValue>) -> Self {
+        #![allow(clippy::mutable_key_type)]
+        let mut headers = hashmap!();
+        for (header_name, header_value) in hyper_headers {
+            headers
+                .entry(header_name)
+                .or_insert_with(Vec::new)
+                .push(header_value.as_bytes().to_vec())
+        }
+        let headers = headers
+            .iter()
+            .map(|item| {
+                (
+                    item.0.to_string(),
+                    HeaderValue {
+                        values: item.1.to_owned(),
+                    },
+                )
+            })
+            .collect();
+
+        OakHeaderMap { headers }
+    }
+}
+
+// Convert String to &'static str
+fn string_to_static_str(s: String) -> &'static str {
+    Box::leak(s.into_boxed_str())
+}
+
+// Create an instance of Oak HttpRequest from the given hyper Request.
+async fn to_oak_http_request(req: Request<Body>) -> HttpRequest {
+    let uri = req.uri().to_string();
+    let method = req.method().as_str().to_string();
+    let headers = OakHeaderMap::from(req.headers()).headers;
+    let body = hyper::body::to_bytes(req.into_body())
+        .await
+        .expect("Error when reading request body from the connection")
```

Comment:
Just checking: is it fine to panic if this happens? Can it not be triggered by a malformed request or something like that?